### PR TITLE
Add a new checker to check cross-feature dependency. 

### PR DESCRIPTION
**Pull request by yutongzhang-microsoft**
### Description of PR

We introduced a new approach to PR testing called `Impacted area based PR testing`. In this model, the scope of PR testing is determined by the specific areas of the code that are impacted by the changes, allowing for more focused and efficient testing.
This means, we need to establish clear boundaries between different sections of code and minimize dependencies as much as possible.

In the tests directory of the `sonic-mgmt` repository, we have categorized scripts into two main groups: shared scripts and feature-specific scripts. Shared scripts provide common utilities or functionality used across multiple features, while feature-specific scripts are tied to particular features and their corresponding logic.

However, the previous codebase contained a significant number of cross-feature dependencies, where scripts from one feature directly referenced or relied on scripts from another. To address this issue and align with our new testing model, we manually reviewed the existing code and removed all cross-feature references. But we need a mechanism to check future modifications and new code to prevent reintroducing these issues.

In this PR, we introduce a new checker to identify any cross-feature dependencies. At this stage, since some cross-feature dependencies remain in the code, this checker is configured to flag these dependencies without causing the entire test to fail. Once all current dependencies are fully removed, any reintroduced cross-feature dependencies detected by this checker will result in a test failure.

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

#### What is the motivation for this PR?
We introduced a new approach to PR testing called `Impacted area based PR testing`. In this model, the scope of PR testing is determined by the specific areas of the code that are impacted by the changes, allowing for more focused and efficient testing.
This means, we need to establish clear boundaries between different sections of code and minimize dependencies as much as possible.

In the tests directory of the `sonic-mgmt` repository, we have categorized scripts into two main groups: shared scripts and feature-specific scripts. Shared scripts provide common utilities or functionality used across multiple features, while feature-specific scripts are tied to particular features and their corresponding logic.

However, the previous codebase contained a significant number of cross-feature dependencies, where scripts from one feature directly referenced or relied on scripts from another. To address this issue and align with our new testing model, we manually reviewed the existing code and removed all cross-feature references. But we need a mechanism to check future modifications and new code to prevent reintroducing these issues.

In this PR, we introduce a new checker to identify any cross-feature dependencies. At this stage, since some cross-feature dependencies remain in the code, this checker is configured to flag these dependencies without causing the entire test to fail. Once all current dependencies are fully removed, any reintroduced cross-feature dependencies detected by this checker will result in a test failure.

#### How did you do it?
In this PR, we introduce a new checker to identify any cross-feature dependencies. At this stage, since some cross-feature dependencies remain in the code, this checker is configured to flag these dependencies without causing the entire test to fail. Once all current dependencies are fully removed, any reintroduced cross-feature dependencies detected by this checker will result in a test failure.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -43,6 +43,20 @@ stages:
       parameters:
         MGMT_BRANCH: ""
 
+  - job: dependency_check
+    displayName: "Dependency Check"
+    timeoutInMinutes: 10
+    continueOnError: true
+    pool: sonic-common
+    steps:
+      - script: |
```
Comment (xwjiang-ms) [inline azure-pipelines.yml:52]: suggest to use a template yml here to make azure-pipelines.yml clean

File: .azure-pipelines/dependency_check/dependency_check.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            imported_module_feature_path = get_feature_path(imported_module["module_path"])
+            if imported_module_feature_path is not None:
+                project_path = os.path.dirname(file_feature_path)
+                # Import from these paths are allowed.
+                if imported_module_feature_path not in [os.path.join(project_path, "common"),
+                                                        os.path.join(project_path, "ptf_runner.py"),
+                                                        os.path.join(project_path, "conftest.py"),
+                                                        file_feature_path]:
+                    print("There is a cross-feature dependence. File: {}, import module: {}"
+                          .format(file_path, imported_module["module"]))
```
Comment (xwjiang-ms) [inline .azure-pipelines/dependency_check/dependency_check.py:177]: Here may have duplicate prints, how about just print 1 line and continue, or you can also print the alias

File: azure-pipelines.yml (hunk 1 above)
Comment (yutongzhang-microsoft) [reply azure-pipelines.yml:52]: There is only one step here, so I think there is no necessary to create another template file here?

File: .azure-pipelines/dependency_check/dependency_check.py (hunk 2 above)
Comment (yutongzhang-microsoft) [reply .azure-pipelines/dependency_check/dependency_check.py:177]: I think one script one line is more clearly for user to read.

File: .azure-pipelines/dependency_check/README.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+```
+ImportFrom(
+    module='x',
+    names=[
+        alias(name='y', asname=None),
+        alias(name='z', asname=None)],
+    level=0)
+```
+
+To achieve our goal, we need to follow these steps. \
```
Comment (wangxin) [inline .azure-pipelines/dependency_check/README.md:61]: Remove extra "\" at the end of line.

File: .azure-pipelines/dependency_check/dependency_check.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import sys
+import os
+import importlib.util
+from natsort import natsorted
+from contextlib import contextmanager
+
+
+def collect_all_scripts():
+    """
+        Recursively find all files ending with ".py" under the folder "tests"
```
[the remark this answers is not in the document]
Comment (yutongzhang-microsoft) [reply .azure-pipelines/dependency_check/dependency_check.py]: I have reformatted these docstrings.

File: .azure-pipelines/dependency_check/dependency_check.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    files = natsorted(files)
+    return files
+
+
+@contextmanager
+def set_sys_path(file_path):
+    """
+        Add the project path into sys path
+    """
+    original_sys_path = sys.path.copy()
```
[the remark this answers is not in the document]
Comment (yutongzhang-microsoft) [reply .azure-pipelines/dependency_check/dependency_check.py:37]: We not only need add the path of the project into sys path, but also need to add the path of the specific file into the sys path. As different files have different path, we should use this function when loop to each file.

File: .azure-pipelines/dependency_check/README.md (hunk 3 above)
Comment (yutongzhang-microsoft) [reply .azure-pipelines/dependency_check/README.md]: done

File: azure-pipelines.yml (hunk 1 above)
Comment (xwjiang-ms) [reply azure-pipelines.yml:52]: I mean if you need to do some change (like pip install, add library...), you can change in the template file but not azure-pipelines.yml, azure-pipelines.yml file should be more static

Comment (yutongzhang-microsoft) [reply azure-pipelines.yml:52]: This makes sence, I will change

**Review by xwjiang-ms — APPROVED** [review]

File: .azure-pipelines/dependency_check/dependency_check.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wangxin) [reply .azure-pipelines/dependency_check/dependency_check.py:37]: But the docstring says "Add the project path into sys path".

[the remark this answers is not in the document]
Comment (yutongzhang-microsoft) [reply .azure-pipelines/dependency_check/dependency_check.py:37]: Modified the docsting.

File: .azure-pipelines/dependency_check/dependency_check.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        file_path (list): A list of files ending with ".py" under the folder "tests"
+    Returns:
+        None
+    """
+    original_sys_path = sys.path.copy()
+    try:
+        current_dir = os.path.abspath(os.path.dirname(file_path))
+        while current_dir != os.path.dirname(current_dir):
+            if current_dir.endswith("/tests"):
+                sys.path.insert(len(sys.path), os.path.join(current_dir, "common"))
```
[the remark this answers is not in the document]
Comment (yutongzhang-microsoft) [reply .azure-pipelines/dependency_check/dependency_check.py]: Modified.

**Review by wangxin — APPROVED** [review]

**Pull request merged**